sugondat/chain/runtimes/sugondat-kusama/tests/fee_adjustments.rs

```diff
@@ -0,0 +1,574 @@
+    // s* = 0.1875 (TargetBlockFullness) or 0.16 (TargetBlockSize)
+    // The bound from the research in an empty chain is:
+    // v <~ (p / k(0 - s*))
+    // p > v * k * -s*
+    // to get p == -1 we'd need
+    // -1 > v * k * -s*
+    // 1 < v * k * s*
+    // 1 / (v * s*) < k
+    //
+    // if s* = 0.1875
+    //  then k > 71_111 ~ 9.8 days
+    // else s* = 0.16
+    //  then k > 83_333 ~ 11.5 days
+
+    let test = |mul_type: MultiplierType| {
+        mul_type.run_with(Weight::zero(), || {
+            // start from 1, the default.
+            let mut fm = Multiplier::one();
+            let mut iterations: u64 = 0;
+            let limit = mul_type.k_bound();
+            loop {
+                let next = mul_type.runtime_multiplier_update(fm);
+                fm = next;
+                if fm <= mul_type.min_multiplier() {
+                    break;
+                }
+                iterations += 1;
+            }
+            assert!(iterations > limit);
+        })
+    };
+
+    test(MultiplierType::Fee);
+    test(MultiplierType::Length);
+}
+
+#[test]
+fn min_change_per_day() {
+    let test = |mul_type: MultiplierType| {
+        mul_type.run_with(mul_type.max(), || {
+            let mut fm = Multiplier::one();
+            // See the example in the doc of `TargetedFeeAdjustment`. are at least 0.234, hence
+            // `fm > 1.234`.
+            for _ in 0..DAYS {
+                let next = mul_type.runtime_multiplier_update(fm);
+                fm = next;
+            }
+            // 7200 blocks per day with one 12 seconds blocks
+            // v * k * (1 - s)
+            let expected = mul_type.adjustment_variable()
+                * Multiplier::saturating_from_integer(7200)
+                * (Multiplier::saturating_from_integer(1) - mul_type.target_percentage());
+            assert!(fm > expected);
+        })
+    };
+    test(MultiplierType::Fee);
+    test(MultiplierType::Length);
+}
+
+#[test]
+#[ignore]
+fn congested_chain_fee_simulation() {
+    // `cargo test congested_chain_simulation -- --nocapture` to get some insight.
+
+    // almost full. The entire quota of normal transactions is taken.
+    let block_weight = BlockWeights::get()
+        .get(DispatchClass::Normal)
+        .max_total
+        .unwrap()
+        - Weight::from_parts(100, 0);
+
+    // Default substrate weight.
+    let tx_weight = frame_support::weights::constants::ExtrinsicBaseWeight::get();
+
+    MultiplierType::Fee.run_with(block_weight, || {
+        // initial value configured on module
+        let mut fm = Multiplier::one();
+        assert_eq!(fm, TransactionPayment::next_fee_multiplier());
+
+        let mut iterations: u64 = 0;
+        loop {
+            let next = MultiplierType::Fee.runtime_multiplier_update(fm);
+            // if no change, panic. This should never happen in this case.
+            if fm == next {
+                panic!("The fee should ever increase");
```

Comment:
never?

Reply:
Yes, I also think it should be always, but I found that comment in the polkadot-sdk repo so I thought It was just me not knowing English enough 😅

I will correct it!